File: rb/human_readable_json_protocol.rb
require 'thrift'
require 'json'

class HumanReadableJsonProtocol < Thrift::BaseProtocol
  METHOD_KEY         = "method"
  SERVICES_KEY       = "services"
  NAME_KEY           = "name"
  ARGUMENTS_KEY      = "arguments"
  RESULT_KEY         = "result"
  SUCCESS_KEY        = "success"
  EXCEPTIONS_KEY     = "exceptions"
  EXCEPTION_KEY      = "exception"
  FUNCTIONS_KEY      = "functions"
  KEY_KEY            = "key"
  ONEWAY_KEY         = "oneway"
  FIELDS_KEY         = "fields"
  STRUCTS_KEY        = "structs"
  MESSAGE_KEY        = "message"
  CLASS_KEY          = "class"
  TYPE_ID_KEY        = "typeId"
  TYPE_KEY           = "type"
  KEY_TYPE_ID_KEY    = "keyTypeId"
  KEY_TYPE_KEY       = "keyType"
  VALUE_TYPE_ID_KEY  = "valueTypeId"
  VALUE_TYPE_KEY     = "valueType"
  ELEM_TYPE_ID_KEY   = "elemTypeId"
  ELEM_TYPE_KEY      = "elemType"
  RETURN_TYPE_ID_KEY = "returnTypeId"
  RETURN_TYPE_KEY    = "returnType"

  class HumanReadableJsonProtocolFactory < Thrift::BaseProtocolFactory
    def initialize(dir)
      @packages_metadata = []
      Dir.foreach(dir) do |fname|
        if fname.end_with? ".json"
          @packages_metadata << JSON.parse(File.read("#{dir}/#{fname}"))
        end
      end
    end

    def get_protocol(trans, service, debug = false)
      HumanReadableJsonProtocol.new(trans, service, @packages_metadata, debug)
    end
  end

  attr_accessor :metadata

  def initialize(protocol, service, packages_metadata, debug = false)
    super(protocol)
    @oprot = Thrift::JsonProtocol.new(protocol)
    @metadata = packages_metadata
    @service = service
    @params = []
    @err = nil
    @debug = debug
    @tabs = 0
  end

  def tab_up
    @tabs = @tabs + 1
  end

  def tab_down
    @tabs = @tabs - 1
  end

  def tabs
    "\t" * @tabs
  end

  def find_in_json_array(arr, key, value)
    arr.find do |e|
      e[key] == value
    end
  end

  def get_method_info(service_name, method_name)
    metadata.each do |v|
      if v[SERVICES_KEY]
        svc = find_in_json_array(v[SERVICES_KEY], NAME_KEY, "#{v[NAME_KEY]}.#{service_name}")
        if svc
          method_info = find_in_json_array(svc[FUNCTIONS_KEY], NAME_KEY, method_name)
          return method_info if method_info
        end
      end
    end
  end

  def get_message_type_and_seq(request, method_info)
    if request.has_key? ARGUMENTS_KEY
      if method_info && method_info.has_key?(ONEWAY_KEY)
        [Thrift::MessageTypes::ONEWAY, 0]
      else
        [Thrift::MessageTypes::CALL, 0]
      end
    elsif request.has_key? RESULT_KEY
      [Thrift::MessageTypes::REPLY, 1]
    elsif request.has_key? EXCEPTION_KEY
      [Thrift::MessageTypes::EXCEPTION, 1]
    else
      raise Thrift::ProtocolException.new(Thrift::ProtocolException::INVALID_DATA, "Unable to parse message type")
    end
  end

  def add(*vals)
    @params = @params.concat(vals)
  end

  def get_info(clazz)
    parts = clazz.split(".")
    program = find_in_json_array(metadata, NAME_KEY, parts[0])
    find_in_json_array(program[STRUCTS_KEY], NAME_KEY, parts[1])
  end

  def get_struct_field_list(elem_type)
    get_info(elem_type[CLASS_KEY])[FIELDS_KEY]
  end

  def raise_expected(type)
    raise Exception.new("Expected #{type}")
  end

  def parse(field_info, value, field_type_id_key, field_type_key)
    field_type = field_info[field_type_id_key]

    case field_type
    when "bool"
      raise_expected("bool") unless value == true || value == false
      add(value)
    when "i8", "i16", "i32", "i64", "double"
      raise_expected("i8") unless value.is_a? Numeric
      add(value)
    when "string"
      raise_expected("string") unless value.is_a? String
      add(value)
    when "struct", "union", "exception"
      parse_struct(get_struct_field_list(field_info[field_type_key]), value)
    when "map"
      parse_map(field_info[field_type_key], value)
    when "set", "list"
      parse_list(field_info[field_type_key], value)
    else
      raise Exception.new("Unexpected type #{field_type}")
    end
  end

  def parse_map(field_info, request)
    raise_expected("JSON Object") unless request.is_a? Hash
    key_type = string_to_type_id(field_info[KEY_TYPE_ID_KEY])
    value_type = string_to_type_id(field_info[VALUE_TYPE_ID_KEY])
    add(key_type, value_type, request.size)
    request.each do |key, value|
      parse(field_info, key, KEY_TYPE_ID_KEY, KEY_TYPE_KEY)
      parse(field_info, value, VALUE_TYPE_ID_KEY, VALUE_TYPE_KEY)
    end
  end

  def parse_list(field_info, request)
    raise_expected("JSON Array") unless request.is_a? Array
    elem_type = string_to_type_id(field_info[ELEM_TYPE_ID_KEY])
    add(elem_type, request.size)
    request.each do |value|
      parse(field_info, value, ELEM_TYPE_ID_KEY, ELEM_TYPE_KEY)
    end
  end

  def parse_struct(fields_list, request)
    raise_expected("JSON Object") unless request.is_a? Hash
    request.each do |key, value|
      field_info = find_in_json_array(fields_list, NAME_KEY, key)
      raise Exception.new("Unexpected key #{key}") unless field_info
      field_type = string_to_type_id(field_info[TYPE_ID_KEY])
      add(key, field_type, field_info[KEY_KEY] || 0)
      parse(field_info, value, TYPE_ID_KEY, TYPE_KEY)
    end
    add("", Thrift::Types::STOP, -1)
  end

  def string_to_type_id(field_type)
    case field_type
    when "bool"
      Thrift::Types::BOOL
    when "i8"
      Thrift::Types::BYTE
    when "i16"
      Thrift::Types::I16
    when "i32"
      Thrift::Types::I32
    when "i64"
      Thrift::Types::I64
    when "double"
      Thrift::Types::DOUBLE
    when "string"
      Thrift::Types::STRING
    when "struct", "union", "exception"
      Thrift::Types::STRUCT
    when "map"
      Thrift::Types::MAP
    when "set"
      Thrift::Types::SET
    when "list"
      Thrift::Types::LIST
    else
      raise Exception.new("Unknown type identifier #{field_type}")
    end
  end

  def read_message_begin
    request = JSON.parse(trans.read(nil))

    name = request[METHOD_KEY]
    method_info = get_method_info(@service, name)
    type_id, seq_id = get_message_type_and_seq(request, method_info)

    if request.has_key? ARGUMENTS_KEY
      unless method_info
        add("", Thrift::Types::STOP, -1)
        return [name, type_id, seq_id]
      end

      parse_struct(method_info[ARGUMENTS_KEY], request[ARGUMENTS_KEY])
    elsif request.has_key? RESULT_KEY
      unless method_info
        add("", Thrift::Types::STOP, -1)
        return [name, type_id, seq_id]
      end

      result = request[RESULT_KEY]
      if result.has_key? SUCCESS_KEY
        begin
          return_type = string_to_type_id(method_info[RETURN_TYPE_ID_KEY])
          add("", return_type, 0)
          parse(method_info, result[SUCCESS_KEY], RETURN_TYPE_ID_KEY, RETURN_TYPE_KEY)
          add("", Thrift::Types::STOP, -1)
        rescue Exception => e
          @err = e
        end
      elsif result.size == 0
        add("", Thrift::Types::STOP, -1)
      else
        err_name = result.keys.first
        err_info = find_in_json_array(method_info[EXCEPTIONS_KEY], NAME_KEY, err_name)
        raise Thrift::ProtocolException.new(Thrift::ProtocolException::INVALID_DATA, "Unable to parse result") unless err_info
        add(err_name, Thrift::Types::STRUCT, err_info[KEY_KEY])
        begin
          parse(err_info, result[err_name], TYPE_ID_KEY, TYPE_KEY)
        rescue Exception => e
          @err = e
        end
        add("", Thrift::Types::STOP, -1)
      end
    elsif request.has_key? EXCEPTION_KEY
      add("", Thrift::Types::STRING, 1)
      add(request[EXCEPTION_KEY][MESSAGE_KEY] || "")
      add("", Thrift::Types::I32, 2)
      add(request[EXCEPTION_KEY][TYPE_KEY] || Thrift::ApplicationException::UNKNOWN)
      add("", Thrift::Types::STOP, -1)
    else
      raise Thrift::ProtocolException.new(Thrift::ProtocolException::INVALID_DATA, "Unable to parse data")
    end
    puts request.inspect if @debug
    puts @params.inspect if @debug
    [name, type_id, seq_id]
  end

  def read_message_end
    # No-op
  end

  def read_struct_begin
    puts "#{tabs}Struct begin" if @debug
    tab_up
    raise @err if @err
  end

  def read_struct_end
    tab_down
    puts "#{tabs}Struct end -- #{@params.size}" if @debug
    # No-op
  end

  def read_field_begin
    puts "#{tabs}Field begin #{@params[0...3]}" if @debug
    tab_up unless @params[1] == Thrift::Types::STOP
    @params.slice!(0, 3)
  end

  def read_field_end
    tab_down
    puts "#{tabs}Field end" if @debug
    # No-op
  end

  def read_map_begin
    puts "#{tabs}Map begin #{@params[0...3]}" if @debug
    tab_up
    @params.slice!(0, 3)
  end

  def read_map_end
    tab_down
    puts "#{tabs}Map end" if @debug
    # No-op
  end

  def read_list_begin
    puts "#{tabs}List begin #{@params[0...2]}" if @debug
    tab_up
    @params.slice!(0, 2)
  end

  def read_list_end
    tab_down
    puts "#{tabs}List end" if @debug
    # No-op
  end

  def read_set_begin
    puts "#{tabs}Set begin #{@params[0...3]}" if @debug
    tab_up
    @params.slice!(0, 2)
  end

  def read_set_end
    tab_down
    puts "#{tabs}Set End" if @debug
    # No-op
  end

  def read_bool
    puts "#{tabs}Bool #{@params[0]}" if @debug
    @params.slice!(0, 1).first
  end

  def read_byte
    puts "#{tabs}Byte #{@params[0]}" if @debug
    @params.slice!(0, 1).first
  end

  def read_i16
    puts "#{tabs}i16 #{@params[0]}" if @debug
    @params.slice!(0, 1).first
  end

  def read_i32
    puts "#{tabs}i32 #{@params[0]}" if @debug
    @params.slice!(0, 1).first
  end

  def read_i64
    puts "#{tabs}i64 #{@params[0]}" if @debug
    @params.slice!(0, 1).first
  end

  def read_double
    puts "#{tabs}Double #{@params[0]}" if @debug
    @params.slice!(0, 1).first
  end

  def read_string
    puts "#{tabs}String #{@params[0]}" if @debug
    @params.slice!(0, 1).first
  end

  def read_binary
    read_string
  end

  def write_message_begin(name, type, seqid)
    @oprot.write_json_object_start
    @oprot.write_string(METHOD_KEY)
    @oprot.write_string(name.to_s)
    case type
    when Thrift::MessageTypes::CALL
      @oprot.write_string(ARGUMENTS_KEY)
    when Thrift::MessageTypes::REPLY
      @oprot.write_string(RESULT_KEY)
    when Thrift::MessageTypes::EXCEPTION
      @oprot.write_string(EXCEPTION_KEY)
    end
  end

  def write_message_end
    @oprot.write_json_object_end
  end

  def write_struct_begin(name)
    @oprot.write_json_object_start
  end

  def write_struct_end
    @oprot.write_json_object_end
  end

  def write_field_begin(name, type, id)
    @oprot.write_string(name.to_s)
  end

  def write_field_end
    # No-op
  end

  def write_field_stop
    # No-op
  end

  def write_map_begin(ktype, vtype, size)
    @oprot.write_json_object_start
  end

  def write_map_end
    @oprot.write_json_object_end
  end

  def write_list_begin(etype, size)
    @oprot.write_json_array_start
  end

  def write_list_end
    @oprot.write_json_array_end
  end

  def write_set_begin(etype, size)
    @oprot.write_json_array_start
  end

  def write_set_end
    @oprot.write_json_array_end
  end

  def write_bool(bool)
    @oprot.write_json_integer(bool)
  end

  def write_byte(byte)
    @oprot.write_byte(byte)
  end

  def write_i16(i16)
    @oprot.write_i16(i16)
  end

  def write_i32(i32)
    @oprot.write_i32(i32)
  end

  def write_i64(i64)
    @oprot.write_i64(i64)
  end

  def write_double(dub)
    @oprot.write_double(dub.to_f)
  end

  def write_string(str)
    @oprot.write_string(str)
  end

  def write_binary(buf)
    @oprot.write_binary(buf)
  end
end

File: java/humanthrift/src/main/java/com/devansh/humanthrift/HumanReadableJsonProtocol.java
package com.devansh.humanthrift;


import org.apache.thrift.TException;
import org.apache.thrift.protocol.*;
import org.apache.thrift.transport.TTransport;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Iterator;
import java.util.LinkedList;

public class HumanReadableJsonProtocol extends TProtocol {

    public static class Factory implements TProtocolFactory {

        private JSONArray metadata;
        private String service;

        public Factory(JSONArray metadata, String service) {
            this.metadata = metadata;
            this.service = service;
        }

        @Override
        public TProtocol getProtocol(TTransport transport) {
            return HumanReadableJsonProtocol.forService(transport, metadata, service);
        }
    }

    private static final String METHOD_KEY = "method";
    private static final String SERVICES_KEY = "services";
    private static final String NAME_KEY = "name";
    private static final String ARGUMENTS_REQUEST_KEY = "arguments";
    private static final String ARGUMENTS_KEY = "arguments";
    private static final String RESULT_KEY = "result";
    private static final String SUCCESS_KEY = "success";
    private static final String EXCEPTIONS_KEY = "exceptions";
    private static final String EXCEPTION_KEY = "exception";
    private static final String FUNCTIONS_KEY = "functions";
    private static final String KEY_KEY = "key";
    private static final String ONEWAY_KEY = "oneway";
    private static final String FIELDS_KEY = "fields";
    private static final String STRUCTS_KEY = "structs";
    private static final String MESSAGE_KEY = "message";
    private static final String CLASS_KEY = "class";
    private static final String TYPE_ID_KEY = "typeId";
    private static final String TYPE_KEY = "type";
    private static final String KEY_TYPE_ID_KEY = "keyTypeId";
    private static final String KEY_TYPE_KEY = "keyType";
    private static final String VALUE_TYPE_ID_KEY = "valueTypeId";
    private static final String VALUE_TYPE_KEY = "valueType";
    private static final String ELEM_TYPE_ID_KEY = "elemTypeId";
    private static final String ELEM_TYPE_KEY = "elemType";
    private static final String RETURN_TYPE_ID_KEY = "returnTypeId";
    private static final String RETURN_TYPE_KEY = "returnType";

    private final JSONArray metadata;
    private final String service;
    private final String struct;
    private final LinkedList<Object> params;
    private final TSimpleJSONProtocol oprot;
    private TException err;
    private boolean structRead;

    private HumanReadableJsonProtocol(TTransport transport, JSONArray metadata, String service, String struct) {
        super(transport);
        this.metadata = metadata;
        this.service = service;
        this.struct = struct;
        this.params = new LinkedList<>();
        oprot = new TSimpleJSONProtocol(transport);
    }

    /**
     * Use this when you want to parse an entire request coming in. You need to pass in the name of the
     * service without the name of the package. You will then be able to use whatever handler you
     * passed in to handle the incoming request
     *
     * @param transport
     * @param metadata
     * @param service
     * @return
     */
    public static HumanReadableJsonProtocol forService(TTransport transport, JSONArray metadata, String service) {
        return new HumanReadableJsonProtocol(transport, metadata, service, null);
    }

    /**
     * Use this when you want to Marhsal/Unmarshal a struct. Make sure the struct name HAS the name of the
     * package. So "package.Struct". The reason for this is that we need to know the package to qucikly find
     * the struct.
     *
     * @param transport
     * @param metadata
     * @param struct
     * @return
     */
    public static HumanReadableJsonProtocol forStruct(TTransport transport, JSONArray metadata, String struct) {
        return new HumanReadableJsonProtocol(transport, metadata, null, struct);
    }

    private JSONObject findInJsonArray(JSONArray arr, String key, String value) throws JSONException {
        for (int i = 0; i < arr.length(); i++) {
            JSONObject obj = arr.getJSONObject(i);
            if (value.equals(obj.getString(key))) {
                return obj;
            }
        }

        return null;
    }

    private JSONObject getMethodInfo(String serviceName, String methodName) throws JSONException {
        for (int i = 0; i < metadata.length(); i++) {
            JSONObject v = metadata.getJSONObject(i);
            if (v.has(SERVICES_KEY)) {
                JSONObject svc = findInJsonArray(v.getJSONArray(SERVICES_KEY), NAME_KEY,
                        v.getString(NAME_KEY) + "." + serviceName);
                if (svc != null) {
                    JSONObject methodInfo =
                            findInJsonArray(svc.getJSONArray(FUNCTIONS_KEY), NAME_KEY, methodName);
                    if (methodInfo != null) {
                        return methodInfo;
                    }
                }
            }
        }

        return null;
    }

    private byte[] getMessageTypeAndSeq(JSONObject request,
                                        JSONObject methodInfo) throws TProtocolException {
        if (request.has(ARGUMENTS_REQUEST_KEY)) {
            if (methodInfo != null && methodInfo.has((ONEWAY_KEY))) {
                return new byte[]{TMessageType.ONEWAY, 0};
            } else {
                return new byte[]{TMessageType.CALL, 0};
            }
        } else if (request.has(RESULT_KEY)) {
            return new byte[]{TMessageType.REPLY, 1};
        } else if (request.has(EXCEPTION_KEY)) {
            return new byte[]{TMessageType.EXCEPTION, 1};
        } else {
            throw new TProtocolException(TProtocolException.INVALID_DATA,
                    new Exception("Unable to parse message type"));
        }
    }

    private void add(Object... vals) {
        for (Object o : vals) {
            params.add(o);
        }
    }

    private JSONObject getInfo(String clazz) throws JSONException {
        String[] parts = clazz.split("\\.");
        JSONObject program = findInJsonArray(metadata, NAME_KEY, parts[0]);
        return findInJsonArray(program.getJSONArray(STRUCTS_KEY), NAME_KEY, parts[1]);
    }

    private JSONArray getStructFieldList(JSONObject elemType) throws JSONException {
        return getInfo(elemType.getString(CLASS_KEY)).getJSONArray(FIELDS_KEY);
    }

    private void raiseExpected(String type, Object got) throws TProtocolException {
        throw new TProtocolException(TProtocolException.INVALID_DATA, new Exception(
                "Expected " + type + " got " +
                        (got == null ? "null" : got.getClass().getSimpleName())));
    }

    private void parse(JSONObject fieldInfo, Object value, String fieldTypeIdKey,
                       String fieldTypeKey) throws JSONException, TProtocolException {
        String fieldType = fieldInfo.getString(fieldTypeIdKey);

        switch (fieldType) {
            case "bool":
                if (value instanceof Boolean) {
                    add(value);
                } else {
                    raiseExpected("bool", value);
                }
                break;

            case "i8":
            case "i16":
            case "i32":
            case "i64":
            case "double":
                if (value instanceof Number) {
                    add(value);
                } else {
                    raiseExpected(fieldType, value);
                }
                break;

            case "string":
                if (value instanceof String) {
                    add(value);
                } else {
                    raiseExpected("string", value);
                }
                break;

            case "struct":
            case "union":
            case "exception":
                parseStruct(getStructFieldList(fieldInfo.getJSONObject(fieldTypeKey)), value);
                break;

            case "map":
                parseMap(fieldInfo.getJSONObject(fieldTypeKey), value);
                break;

            case "set":
                parseSet(fieldInfo.getJSONObject(fieldTypeKey), value);
                break;
            case "list":
                parseList(fieldInfo.getJSONObject(fieldTypeKey), value);
                break;

            default:
                throw new TProtocolException(TProtocolException.INVALID_DATA,
                        new Exception("Unexpected type " + fieldType));
        }
    }

    private void parseMap(JSONObject fieldInfo,
                          Object request) throws TProtocolException, JSONException {
        if (!(request instanceof JSONObject)) {
            raiseExpected("JSON Object", request);
        }
        JSONObject jsonObject = (JSONObject) request;

        byte keyType = stringToTypeId(fieldInfo.getString(KEY_TYPE_ID_KEY));
        byte valueType = stringToTypeId(fieldInfo.getString(VALUE_TYPE_ID_KEY));
        add(keyType, valueType, jsonObject.length());

        Iterator<String> it = jsonObject.keys();
        while (it.hasNext()) {
            String key = it.next();
            Object value = jsonObject.get(key);
            parse(fieldInfo, key, KEY_TYPE_ID_KEY, KEY_TYPE_KEY);
            parse(fieldInfo, value, VALUE_TYPE_ID_KEY, VALUE_TYPE_KEY);
        }
    }

    private void parseList(JSONObject fieldInfo,
                           Object request) throws TProtocolException, JSONException {
        if (!(request instanceof JSONArray)) {
            raiseExpected("JSON Array", request);
        }
        JSONArray jsonArray = (JSONArray) request;

        byte elemType = stringToTypeId(fieldInfo.getString(ELEM_TYPE_ID_KEY));
        add(elemType, jsonArray.length());

        for (int i = 0; i < jsonArray.length(); i++) {
            parse(fieldInfo, jsonArray.get(i), ELEM_TYPE_ID_KEY, ELEM_TYPE_KEY);
        }
    }

    private void parseSet(JSONObject fieldInfo,
                          Object request) throws TProtocolException, JSONException {
        Iterator<? extends Object> it = null;
        int size = 0;
        if (request instanceof JSONObject) {
            it = ((JSONObject) request).keys();
            size = ((JSONObject) request).length();
        } else if (request instanceof JSONArray) {
            it = ((JSONArray) request).iterator();
            size = ((JSONArray) request).length();
        } else {
            raiseExpected("JSON Array", request);
        }

        byte elemType = stringToTypeId(fieldInfo.getString(ELEM_TYPE_ID_KEY));
        add(elemType, size);

        while (it.hasNext()) {
            Object val = it.next();
            if (val instanceof String && !fieldInfo.getString(ELEM_TYPE_ID_KEY).equals("string")) {
                val = Integer.parseInt((String) val);
            }
            parse(fieldInfo, val, ELEM_TYPE_ID_KEY, ELEM_TYPE_KEY);
        }
    }

    public void parseStruct(JSONArray fieldsList,
                            Object request) throws TProtocolException, JSONException {
        if (!(request instanceof JSONObject)) {
            raiseExpected("JSON Object", request);
        }
        JSONObject jsonObject = (JSONObject) request;

        Iterator<String> it = jsonObject.keys();
        while (it.hasNext()) {
            String key = it.next();
            Object value = jsonObject.get(key);
            JSONObject fieldInfo = findInJsonArray(fieldsList, NAME_KEY, key);
            if (fieldInfo == null) {
                throw new TProtocolException(TProtocolException.INVALID_DATA,
                        new Exception("Unexpected key " + key));
            }

            byte fieldType = stringToTypeId(fieldInfo.getString(TYPE_ID_KEY));
            add(key, fieldType, (short) fieldInfo.optInt(KEY_KEY, 0));
            parse(fieldInfo, value, TYPE_ID_KEY, TYPE_KEY);
        }
        add("", TType.STOP, (short) -1);
    }

    private byte stringToTypeId(String fieldType) throws TProtocolException {
        switch (fieldType) {
            case "bool":
                return TType.BOOL;
            case "i8":
                return TType.BYTE;
            case "i16":
                return TType.I16;
            case "i32":
                return TType.I32;
            case "i64":
                return TType.I64;
            case "double":
                return TType.DOUBLE;
            case "string":
                return TType.STRING;
            case "struct":
            case "union":
            case "exception":
                return TType.STRUCT;
            case "map":
                return TType.MAP;
            case "set":
                return TType.SET;
            case "list":
                return TType.LIST;
            default:
                throw new TProtocolException(TProtocolException.INVALID_DATA,
                        new Exception("Unknown type identifier " + fieldType));
        }
    }

    @Override
    public TMessage readMessageBegin() throws TException {
        try {
            if (service == null) {
                throw new TException("Service Name is required to parse JSON");
            }
            return readMessageBeginHelper();
        } catch (Exception e) {
            throw new TException(e);
        }
    }

    private JSONObject readAllFromTransport() throws JSONException, IOException {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();

        TTransport transport = getTransport();
        byte buffer[] = new byte[1024];
        while (true) {
            try {
                int amt = transport.read(buffer, 0, buffer.length);
                if (amt <= 0) {
                    break;
                }
                byteArrayOutputStream.write(buffer, 0, amt);
            } catch (Exception e) {
                break;
            }
        }

        return new JSONObject(byteArrayOutputStream.toString("UTF-8"));
    }

    private TMessage readMessageBeginHelper() throws JSONException, TProtocolException, IOException {
        JSONObject request = readAllFromTransport();

        String name = request.getString(METHOD_KEY);
        JSONObject methodInfo = getMethodInfo(service, name);
        byte[] messageTypeAndSeq = getMessageTypeAndSeq(request, methodInfo);
        byte typeId = messageTypeAndSeq[0];
        byte seqId = messageTypeAndSeq[1];

        if (request.has(ARGUMENTS_REQUEST_KEY)) {
            if (methodInfo == null) {
                add("", TType.STOP, (short) -1);
                return new TMessage(name, typeId, seqId);
            }

            parseStruct(methodInfo.getJSONArray(ARGUMENTS_KEY), request.get(ARGUMENTS_REQUEST_KEY));
        } else if (request.has(RESULT_KEY)) {
            if (methodInfo == null) {
                add("", TType.STOP, (short) -1);
                return new TMessage(name, typeId, seqId);
            }

            JSONObject result = request.getJSONObject(RESULT_KEY);
            if (result.has(SUCCESS_KEY)) {
                try {
                    byte returnType = stringToTypeId(methodInfo.getString(RETURN_TYPE_ID_KEY));
                    add("", returnType, (short) 0);
                    parse(methodInfo, result.get(SUCCESS_KEY), RETURN_TYPE_ID_KEY, RETURN_TYPE_KEY);
                    add("", TType.STOP, (short) -1);
                } catch (Exception e) {
                    err = new TException(e);
                }
            } else if (result.length() == 0) {
                add("", TType.STOP, (short) -1);
            } else {
                String errName = (String) result.keys().next();
                JSONObject errInfo =
                        findInJsonArray(methodInfo.getJSONArray(EXCEPTIONS_KEY), NAME_KEY, errName);
                if (errInfo == null) {
                    throw new TProtocolException(TProtocolException.INVALID_DATA,
                            new Exception("Unable to parse result"));
                }
                add(errName, TType.STRUCT, (short) errInfo.getInt(KEY_KEY));

                try {
                    parse(errInfo, result.get(errName), TYPE_ID_KEY, TYPE_KEY);
                } catch (Exception e) {
                    err = new TException(e);
                }
                add("", TType.STOP, (short) -1);
            }
        } else if (request.has(EXCEPTION_KEY)) {
            add("", TType.STRING, (short) 1);
            add(request.getJSONObject(EXCEPTION_KEY).optString(MESSAGE_KEY, ""));
            add("", TType.I32, (short) 2);
            add(request.getJSONObject(EXCEPTION_KEY).optInt(TYPE_KEY, TProtocolException.UNKNOWN));
            add("", TType.STOP, (short) -1);
        } else {
            throw new TProtocolException(TProtocolException.INVALID_DATA,
                    new Exception("Unable to parse result"));
        }

        return new TMessage(name, typeId, seqId);
    }

    @Override
    public void readMessageEnd() throws TException {
        //  No-op
    }

    @Override
    public TStruct readStructBegin() throws TException {
        if (err != null) {
            throw err;
        }

        if (service == null && struct != null && !structRead) {
            structRead = true;
            try {
                readStructBeginHelper();
            } catch (Exception e) {
                throw new TException(e);
            }
        }

        return new TStruct();
    }

    private void readStructBeginHelper() throws JSONException, TException, IOException {
        JSONObject request = readAllFromTransport();
        parseStruct(getInfo(struct).getJSONArray(FIELDS_KEY), request);
    }

    @Override
    public void readStructEnd() throws TException {
        //  No-op
    }

    @Override
    public TField readFieldBegin() throws TException {
        return new TField((String) params.pollFirst(), (Byte) params.pollFirst(),
                (Short) params.pollFirst());
    }

    @Override
    public void readFieldEnd() throws TException {
        //  No-op
    }

    @Override
    public TMap readMapBegin() throws TException {
        return new TMap((Byte) params.pollFirst(), (Byte) params.pollFirst(),
                (Integer) params.pollFirst());
    }

    @Override
    public void readMapEnd() throws TException {
        //  No-op
    }

    @Override
    public TList readListBegin() throws TException {
        return new TList((Byte) params.pollFirst(), (Integer) params.pollFirst());
    }

    @Override
    public void readListEnd() throws TException {
        //  No-op
    }

    @Override
    public TSet readSetBegin() throws TException {
        return new TSet((Byte) params.pollFirst(), (Integer) params.pollFirst());
    }

    @Override
    public void readSetEnd() throws TException {
        //  No-op
    }

    @Override
    public boolean readBool() throws TException {
        return (boolean) params.pollFirst();
    }

    @Override
    public byte readByte() throws TException {
        return ((Number) params.pollFirst()).byteValue();
    }

    @Override
    public short readI16() throws TException {
        return ((Number) params.pollFirst()).shortValue();
    }

    @Override
    public int readI32() throws TException {
        return ((Number) params.pollFirst()).intValue();
    }

    @Override
    public long readI64() throws TException {
        return ((Number) params.pollFirst()).longValue();
    }

    @Override
    public double readDouble() throws TException {
        return ((Number) params.pollFirst()).doubleValue();
    }

    @Override
    public String readString() throws TException {
        return (String) params.pollFirst();
    }

    @Override
    public ByteBuffer readBinary() throws TException {
        return ByteBuffer.wrap(readString().getBytes());
    }

    @Override
    public void writeMessageBegin(TMessage tMessage) throws TException {
        oprot.writeStructBegin(null);
        oprot.writeString(METHOD_KEY);
        oprot.writeString(tMessage.name);
        switch (tMessage.type) {
            case TMessageType.CALL:
                oprot.writeString(ARGUMENTS_REQUEST_KEY);
                break;
            case TMessageType.REPLY:
                oprot.writeString(RESULT_KEY);
                break;
            case TMessageType.EXCEPTION:
                oprot.writeString(EXCEPTION_KEY);
                break;
        }
    }

    @Override
    public void writeMessageEnd() throws TException {
        oprot.writeStructEnd();
    }

    @Override
    public void writeStructBegin(TStruct tStruct) throws TException {
        oprot.writeStructBegin(tStruct);
    }

    @Override
    public void writeStructEnd() throws TException {
        oprot.writeStructEnd();
    }

    @Override
    public void writeFieldBegin(TField tField) throws TException {
        oprot.writeString(tField.name);
    }

    @Override
    public void writeFieldEnd() throws TException {
        // No-op
    }

    @Override
    public void writeFieldStop() throws TException {
        // No-op
    }

    @Override
    public void writeMapBegin(TMap tMap) throws TException {
        oprot.writeStructBegin(null);
    }

    @Override
    public void writeMapEnd() throws TException {
        oprot.writeStructEnd();
    }

    @Override
    public void writeListBegin(TList tList) throws TException {
        oprot.writeListBegin(tList);
    }

    @Override
    public void writeListEnd() throws TException {
        oprot.writeListEnd();
    }

    @Override
    public void writeSetBegin(TSet tSet) throws TException {
        oprot.writeSetBegin(tSet);
    }

    @Override
    public void writeSetEnd() throws TException {
        oprot.writeSetEnd();
    }


    @Override
    public void writeBool(boolean b) throws TException {
        oprot.writeBool(b);
    }

    @Override
    public void writeByte(byte b) throws TException {
        oprot.writeByte(b);
    }

    @Override
    public void writeI16(short i) throws TException {
        oprot.writeI16(i);
    }

    @Override
    public void writeI32(int i) throws TException {
        oprot.writeI32(i);
    }

    @Override
    public void writeI64(long l) throws TException {
        oprot.writeI64(l);
    }

    @Override
    public void writeDouble(double v) throws TException {
        oprot.writeDouble(v);
    }

    @Override
    public void writeString(String s) throws TException {
        oprot.writeString(s);
    }

    @Override
    public void writeBinary(ByteBuffer byteBuffer) throws TException {
        oprot.writeBinary(byteBuffer);
    }
}